core/injector/application_injector.cpp

```diff
@@ -991,71 +991,77 @@ namespace {
         di::bind<storage::trie::PolkadotTrieFactory>.template to<storage::trie::PolkadotTrieFactoryImpl>(),
         di::bind<storage::trie::Codec>.template to<storage::trie::PolkadotCodec>(),
         di::bind<storage::trie::TrieSerializer>.template to<storage::trie::TrieSerializerImpl>(),
-        bind_by_lambda<storage::trie_pruner::TriePruner>([](auto const
-                                                                &injector) {
-          auto app_config =
-              injector.template create<sptr<application::AppConfiguration>>();
-          if (!app_config->statePruningDepth().has_value()) {
-            class IdlePruner final : public storage::trie_pruner::TriePruner {
-             public:
-              virtual outcome::result<void> addNewState(
-                  storage::trie::PolkadotTrie const &,
-                  storage::trie::StateVersion) override {
-                return outcome::success();
-              }
-
-              virtual outcome::result<void> addNewChildState(
-                  storage::trie::RootHash const &,
-                  storage::trie::PolkadotTrie const &,
-                  storage::trie::StateVersion) override {
-                return outcome::success();
-              }
-
-              virtual outcome::result<void> markAsChild(Parent,
-                                                        Child) override {
-                return outcome::success();
-              }
-
-              virtual outcome::result<void> pruneFinalized(
-                  primitives::BlockHeader const &,
-                  primitives::BlockInfo const &) override {
-                return outcome::success();
-              }
-
-              virtual outcome::result<void> pruneDiscarded(
-                  primitives::BlockHeader const &) override {
-                return outcome::success();
-              }
-
-              virtual primitives::BlockNumber getBaseBlock() const override {
-                return 0;
-              }
-
-              virtual std::optional<uint32_t> getPruningDepth() const override {
-                return {};
+        bind_by_lambda<storage::trie_pruner::TriePruner>(
```

Comment:
Can we inject it without lambda?

Reply:
I'll try to move the logic from create() to prepare() and then we probably can.